### PR TITLE
[MXNET-1253] fix control_flow_op

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
Support large array in control_flow_ops

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- test_large_array.py:test_where()

### CODE REVIEW

**Comment by apeforest** [comment]
@zheng-da @anirudh2290 @sandeep-krishnamurthy @yuxihu @frankfliu Please review

File: src/operator/tensor/control_flow_op.h [hunk 1]
```diff
@@ -92,7 +92,7 @@ struct where_batch {
   // DType is the output data type
   // CType is the condition data type
   template<typename DType, typename CType>
-  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+  MSHADOW_XINLINE static void Map(index_t i, DType* out, const CType* cond,
                                   const DType* x, const DType* y, int M) {
```
Comment (TaoLv) [inline src/operator/tensor/control_flow_op.h]: `index_t M`?

File: src/operator/tensor/control_flow_op.h [hunk 2]
```diff
@@ -161,7 +161,7 @@ struct where_batch_backward {
   // DType is the output data type
   // CType is condition data type
   template<typename DType, typename CType>
-  MSHADOW_XINLINE static void Map(int i, DType* grad_out,
+  MSHADOW_XINLINE static void Map(index_t i, DType* grad_out,
                                   const DType* grad_in,
                                   const CType* cond, int M) {
```
Comment (TaoLv) [inline src/operator/tensor/control_flow_op.h]: `index_t M`?

File: src/operator/tensor/control_flow_op.h (hunk 1 above)
Comment (apeforest) [reply src/operator/tensor/control_flow_op.h]: Thanks for catching this. Updated

File: src/operator/tensor/control_flow_op.h (hunk 2 above)
Comment (apeforest) [reply src/operator/tensor/control_flow_op.h]: Thanks for catching this. Updated

File: tests/nightly/test_large_array.py [hunk 3]
```diff
@@ -134,6 +134,13 @@ def test_Dense(ctx=mx.cpu(0)):
     res.wait_to_read()
     assert res.shape == (50000000, 100)
 
+def test_where():
+    a = nd.ones(shape=(LARGE_X, SMALL_Y))
+    b = nd.arange(0, LARGE_X).reshape(LARGE_X, 1)
+    b = nd.broadcast_to(b, shape=(b.shape[0], SMALL_Y))
+    res = nd.where(b > 100, a, b)
```
Comment (sandeep-krishnamurthy) [inline tests/nightly/test_large_array.py]: Can we add test for csr data to test where_csr?

Comment (apeforest) [reply tests/nightly/test_large_array.py]: Thanks. test added

**Comment by apeforest** [comment]
@mxnet-label-bot Update [pr-awaiting-merge]

**Comment by apeforest** [comment]
@mxnet-label-bot Update [pr-awaiting-merge, Operator]

**Comment by sandeep-krishnamurthy** [comment]
@TaoLv  - I see your comments are addressed. This is good to merge, I will wait for your confirmation.

**Comment by anirudh2290** [comment]
Can you add the remaining test for `scatter_nd` or is it planned for a different PR ?

**Comment by apeforest** [comment]
@anirudh2290 It's planned for a different PR as this PR does not directly fix `scatter_nd`

File: tests/nightly/test_large_array.py (hunk 3 above)
Comment (anirudh2290) [reply tests/nightly/test_large_array.py]: can we add a test for backward too.

**Comment by TaoLv** [comment]
Seems all comments are addressed. Now merging.

**Pull request merged**